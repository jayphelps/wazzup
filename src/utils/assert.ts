export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
