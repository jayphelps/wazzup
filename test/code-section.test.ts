import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Code SectionÂ (aka function bodies)', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds codes from different functions', () => {
    const type1 = module.addFunctionType([], [wasm.Type.i32]);
    const body1 = [
      new ast.Const(1, wasm.Opcode.I32_CONST)
    ];

    module.addFunction('foo', type1, [], body1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func (result i32)))
      |   (func (;0;) (type 0) (result i32)
      |     i32.const 1))
    `);

    const type2 = module.addFunctionType();
    const body2 = [
      new ast.Const(1, wasm.Opcode.I32_CONST),
      new ast.Drop()
    ];

    module.addFunction('bar', type2, [], body2);

    assertModule(module, `
      | (module
      |   (type (;0;) (func (result i32)))
      |   (type (;1;) (func))
      |   (func (;0;) (type 0) (result i32)
      |     i32.const 1)
      |   (func (;1;) (type 1)
      |     i32.const 1
      |     drop))
    `);
  });

  test('adds drop', () => {
    const type1 = module.addFunctionType([], []);
    const body1 = [
      new ast.Const(1, wasm.Opcode.I32_CONST),
      new ast.Drop()
    ];

    module.addFunction('foo', type1, [], body1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0)
      |     i32.const 1
      |     drop))
    `);
  });

  test('adds constants', () => {
    const type1 = module.addFunctionType([], [wasm.Type.i64]);
    const body1 = [
      new ast.Const(1, wasm.Opcode.I32_CONST),
      new ast.Drop(),
      new ast.Const(2, wasm.Opcode.I64_CONST)
    ];

    module.addFunction('foo', type1, [], body1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func (result i64)))
      |   (func (;0;) (type 0) (result i64)
      |     i32.const 1
      |     drop
      |     i64.const 2))
    `);
  });

  test('adds loads', () => {
    const type1 = module.addFunctionType([], []);
    const body1 = [
      // i32.load8_s
      new ast.Load(
        1,
        true,
        0,
        1,
        wasm.Type.i32,
        new ast.Const(1, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i32.load16_s
      new ast.Load(
        2,
        true,
        1,
        2,
        wasm.Type.i32,
        new ast.Const(2, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i32.load
      new ast.Load(
        4,
        true,
        2,
        3,
        wasm.Type.i32,
        new ast.Const(3, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i64.load8_u
      new ast.Load(
        1,
        false,
        0,
        1,
        wasm.Type.i64,
        new ast.Const(1, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i64.load16_u
      new ast.Load(
        2,
        false,
        1,
        2,
        wasm.Type.i64,
        new ast.Const(2, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i64.load32_u
      new ast.Load(
        4,
        false,
        2,
        3,
        wasm.Type.i64,
        new ast.Const(3, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),

      // i64.load
      new ast.Load(
        8,
        false,
        3,
        4,
        wasm.Type.i64,
        new ast.Const(4, wasm.Opcode.I32_CONST)
      ),
      new ast.Drop(),
    ];

    module.addMemory(1, 1, []);
    module.addFunction('foo', type1, [], body1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0)
      |     i32.const 1
      |     i32.load8_s
      |     drop
      |     i32.const 2
      |     i32.load16_s offset=1
      |     drop
      |     i32.const 3
      |     i32.load offset=2 align=2
      |     drop
      |     i32.const 1
      |     i64.load8_u
      |     drop
      |     i32.const 2
      |     i64.load16_u offset=1
      |     drop
      |     i32.const 3
      |     i64.load32_u offset=2 align=2
      |     drop
      |     i32.const 4
      |     i64.load offset=3 align=4
      |     drop)
      |   (memory (;0;) 1 1))
    `);
  });
});
