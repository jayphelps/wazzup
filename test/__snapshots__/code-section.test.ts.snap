// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Section (aka function bodies) adds codes from different functions 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [
        Const {
          "opcode": 65,
          "returnType": 127,
          "value": 1,
        },
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [
          127,
        ],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Code Section (aka function bodies) adds codes from different functions 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  5,
  1,
  96,
  0,
  1,
  127,
  3,
  2,
  1,
  0,
  10,
  6,
  1,
  4,
  0,
  65,
  1,
  11,
]
`;

exports[`Code Section (aka function bodies) adds codes from different functions 3`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [
        Const {
          "opcode": 65,
          "returnType": 127,
          "value": 1,
        },
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [
          127,
        ],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
    Function {
      "body": Array [
        Const {
          "opcode": 65,
          "returnType": 127,
          "value": 1,
        },
        Drop {},
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "bar",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Code Section (aka function bodies) adds codes from different functions 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  8,
  2,
  96,
  0,
  1,
  127,
  96,
  0,
  0,
  3,
  3,
  2,
  0,
  1,
  10,
  12,
  2,
  4,
  0,
  65,
  1,
  11,
  5,
  0,
  65,
  1,
  26,
  11,
]
`;

exports[`Code Section (aka function bodies) adds constants 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [
        Const {
          "opcode": 65,
          "returnType": 127,
          "value": 1,
        },
        Drop {},
        Const {
          "opcode": 66,
          "returnType": 126,
          "value": 2,
        },
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [
          126,
        ],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [
        126,
      ],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Code Section (aka function bodies) adds constants 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  5,
  1,
  96,
  0,
  1,
  126,
  3,
  2,
  1,
  0,
  10,
  9,
  1,
  7,
  0,
  65,
  1,
  26,
  66,
  2,
  11,
]
`;

exports[`Code Section (aka function bodies) adds drop 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [
        Const {
          "opcode": 65,
          "returnType": 127,
          "value": 1,
        },
        Drop {},
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Code Section (aka function bodies) adds drop 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  10,
  7,
  1,
  5,
  0,
  65,
  1,
  26,
  11,
]
`;

exports[`Code Section (aka function bodies) adds loads 1`] = `
Module {
  "exports": Array [],
  "functions": Array [
    Function {
      "body": Array [
        Load {
          "alignment": 1,
          "byteCount": 1,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 1,
          },
          "offset": 0,
          "signed": true,
          "type": 127,
        },
        Drop {},
        Load {
          "alignment": 2,
          "byteCount": 2,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 2,
          },
          "offset": 1,
          "signed": true,
          "type": 127,
        },
        Drop {},
        Load {
          "alignment": 3,
          "byteCount": 4,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 3,
          },
          "offset": 2,
          "signed": true,
          "type": 127,
        },
        Drop {},
        Load {
          "alignment": 1,
          "byteCount": 1,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 1,
          },
          "offset": 0,
          "signed": false,
          "type": 126,
        },
        Drop {},
        Load {
          "alignment": 2,
          "byteCount": 2,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 2,
          },
          "offset": 1,
          "signed": false,
          "type": 126,
        },
        Drop {},
        Load {
          "alignment": 3,
          "byteCount": 4,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 3,
          },
          "offset": 2,
          "signed": false,
          "type": 126,
        },
        Drop {},
        Load {
          "alignment": 4,
          "byteCount": 8,
          "expression": Const {
            "opcode": 65,
            "returnType": 127,
            "value": 4,
          },
          "offset": 3,
          "signed": false,
          "type": 126,
        },
        Drop {},
      ],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [
    Memory {
      "externalKind": 2,
      "initialSize": 1,
      "maxSize": 1,
      "segments": Array [],
    },
  ],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Code Section (aka function bodies) adds loads 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  5,
  4,
  1,
  1,
  1,
  1,
  10,
  46,
  1,
  44,
  0,
  65,
  1,
  44,
  0,
  0,
  26,
  65,
  2,
  46,
  1,
  1,
  26,
  65,
  3,
  40,
  1,
  2,
  26,
  65,
  1,
  49,
  0,
  0,
  26,
  65,
  2,
  51,
  1,
  1,
  26,
  65,
  3,
  53,
  1,
  2,
  26,
  65,
  4,
  41,
  2,
  3,
  26,
  11,
]
`;
